core/src/smartcontracts/isi/world.rs

```diff
@@ -292,16 +292,19 @@ pub mod isi {
 
     /// Remove all tokens with specified definition id from all accounts in all domains
     fn remove_token_from_accounts(
-        wsv: &WorldStateView,
+        wsv: &mut WorldStateView,
         target_definition_id: &<PermissionTokenDefinition as Identifiable>::Id,
     ) -> Result<(), Error> {
         let mut accounts_with_token = std::collections::HashMap::new();
 
-        for domain in wsv.domains().iter() {
+        let wsv_clone = wsv.clone();
```

Comment:
Cloning the `wsv` is an expensive operation. We need to discuss how to optimise it.

Reply:
I would suggeest the following tweek to avoid clone:
```suggestion
        for domain in wsv.domains().iter() {
            let account_ids = domain.accounts().map(|account| {
                (
                    account.id().clone(),
                    wsv
                        .account_inherent_permission_tokens(account)
                        .filter(|token| token.definition_id() == target_definition_id)
                        .collect::<Vec<_>>(),
                )
            });

            accounts_with_token.extend(account_ids);
        }

        for (account_id, tokens) in accounts_with_token {
            for token in tokens {
                // Most important part move condition out of closure
                if !wsv.remove_account_permission(&account_id, &token) {
                    error!(%token, "token not found - this is a bug");
                }
                wsv.modify_account(&account_id, |account| {
                    let id = account.id();
                    Ok(AccountEvent::PermissionRemoved(id.clone()))
                })?;
            }
        }
```

Reply:
I have a plan for removing nearly all the clones. It's a big change though since we would have to change to a patch based model for updating the world state view. But that way you hold a immutable reference to the world state and accumulate patches as you perform modifications. Then at the end you apply all the patches to the world state. You can also add caching to this model for more speedup.

Reply:
resolved with rollback